pkg/sharings/send_mails.go

```diff
@@ -0,0 +1,222 @@
+	// the description and the sharer's public name.
+	mailValues := &mailTemplateValues{}
+
+	// We get the Couchdb document describing the instance to get the sharer's
+	// public name.
+	doc := &couchdb.JSONDoc{}
+	err := couchdb.GetDoc(instance, consts.Settings,
+		consts.InstanceSettingsID, doc)
+	if err != nil {
+		return err
+	}
+	sharerPublicName, _ := doc.M["public_name"].(string)
+	mailValues.SharerPublicName = sharerPublicName
+
+	// Fill in the description.
+	if s.Desc == "" {
+		mailValues.Description = "[No description provided]"
+	} else {
+		mailValues.Description = s.Desc
+	}
+
+	// For each recipient we send the email and update the status if an error
+	// occurred.
+	recipientsStatus, err := s.RecStatus(instance)
+	if err != nil {
+		return err
+	}
+
+	errorOccurred := false
+	for _, rs := range recipientsStatus {
+		recipient := rs.recipient
+
+		// Generate recipient specific OAuth query string.
+		recipientOAuthQueryString, err := generateOAuthQueryString(s, recipient)
+		if err != nil {
+			errorOccurred = logErrorAndSetRecipientStatus(rs, err)
+			break
+		}
+
+		// Augment base values with recipient specific information.
+		mailValues.RecipientName = recipient.Email
+		mailValues.OAuthQueryString = recipientOAuthQueryString
+
+		sharingMessage, err := generateMailMessage(s, recipient, mailValues)
+		if err != nil {
+			errorOccurred = logErrorAndSetRecipientStatus(rs, err)
+			break
+		}
+
+		// We ask to queue a new mail job.
+		// XXX Do we have to do something with the returned values?
+		_, _, err = instance.JobsBroker().PushJob(&jobs.JobRequest{
+			WorkerType: "sendmail",
+			Options:    nil,
+			Message:    sharingMessage,
+		})
+		if err != nil {
+			errorOccurred = logErrorAndSetRecipientStatus(rs, err)
+			break
+		}
+	}
+
+	if errorOccurred {
+		return ErrMailCouldNotBeSent
+	}
+
+	return nil
+}
+
+// logErrorAndSetRecipientStatus will log an error in the stack and set the
+// status of the impacted recipient to "error".
+func logErrorAndSetRecipientStatus(rs *RecipientStatus, err error) bool {
+	log.Error(`[Sharing] An error occurred while trying to send
+        the email invitation`, err)
+
+	rs.Status = consts.ErrorStatus
+	return true
+}
+
+// generatemailMessage will extract and compute the relevant information
+// from the sharing to generate the mail we will send to the recipient
+// specified.
+func generateMailMessage(s *Sharing, r *Recipient,
+	mailValues *mailTemplateValues) (*jobs.Message, error) {
+
+	// We create the mail parts: its content.
+	mailPartEn := workers.MailPart{
+		Type: "text/html",
+		Body: mailTemplateEn}
+
+	mailPartFr := workers.MailPart{
+		Type: "text/html",
+		Body: mailTemplateFr}
+
+	mailParts := []*workers.MailPart{&mailPartEn, &mailPartFr}
+
+	// The address of the recipient.
+	// XXX In addition to testing if the email was provided, do we test if it
+	// actually corresponds to an email or do we "wait" for the error triggered
+	// by the mail job?
```

Comment:
the mail sender is asynchronous so testing that the mail has been send can take an unknown amount of time.

if the mail are not "reliables", we should test their well formatting here. although the formatting of an email does of course not tell if the mail address is actually a good one.